clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -633,77 +633,94 @@ raw2trace_t::do_conversion()
     return "";
 }
 
-uint
-raw2trace_t::hash_instr_summary(void *key)
+raw2trace_t::block_summary_t *
+raw2trace_t::lookup_block_summary(void *tls, app_pc block_start)
 {
-    // Our key is a payload instance.
-    const instr_summary_t *instr = static_cast<const instr_summary_t *>(key);
-    // We assume tail-duplication is rare.
-    // We target x86 and use all bits of the PC.
-    // hashtable_t masks this off to the size of the table for us.
-    return static_cast<uint>(reinterpret_cast<ptr_uint_t>(instr->pc()));
-}
-
-bool
-raw2trace_t::cmp_instr_summary(void *val1, void *val2)
-{
-    const instr_summary_t *instr1 = static_cast<const instr_summary_t *>(val1);
-    const instr_summary_t *instr2 = static_cast<const instr_summary_t *>(val2);
-    return (instr1->tag() == instr2->tag() && instr1->pc() == instr2->pc());
+    auto tdata = reinterpret_cast<raw2trace_thread_data_t *>(tls);
```

Comment:
This type opacity is part of the interface separation that allows replacing the upstream implementation with an alternative.  The general separation was put in by Mircea if you recall, as part of #3129.  I have been wondering if it would be cleanest to go and remove the separation and use direct types in a single implementation, since nobody is using a different implementation.  But that would be best done separately.  I went ahead and filed #4062 on such a refactoring.

Reply:
I added a comment to the header referencing #4062.